src/coreclr/jit/codegenloongarch64.cpp

```diff
@@ -5667,15 +5674,15 @@ void CodeGen::genCodeForTreeNode(GenTree* treeNode)
 //
 void CodeGen::genSetRegToIcon(regNumber reg, ssize_t val, var_types type)
 {
+    emitAttr attr = emitActualTypeSize(type);
+    if (!compiler->opts.compReloc)
+    {
+        attr = EA_SIZE(attr);
+    }
     // Reg cannot be a FP reg
     assert(!genIsValidFloatReg(reg));
 
-    // The only TYP_REF constant that can come this path is a managed 'null' since it is not
-    // relocatable.  Other ref type constants (e.g. string objects) go through a different
```

Comment:
Why is this no longer the case?

Reply:
（1）The `var_types type` of `CodeGen::genSetRegToIcon` may be  GC type only when is called by `CodeGen::genSetRegToConst`.
（2） The `CodeGen::genSetRegToConst` for arm64 where the `attr` maybe GC type https://github.com/dotnet/runtime/blob/8b19f23d99b71c3152f02163b3b7275d8b24678e/src/coreclr/jit/codegenarm64.cpp#L2322-L2328

But the `CodeGen::instGen_Set_Reg_To_Imm()` for arm64 which liking the following
https://github.com/dotnet/runtime/blob/8b19f23d99b71c3152f02163b3b7275d8b24678e/src/coreclr/jit/codegenarm64.cpp#L2204-L2214
If the `size` is GC type while the `compiler->opts.compReloc` is false, the GC type info will be stripped here.

So for LA64, I just delete this to fixed the assert error.

Reply:
Thanks for the context. I see what you are doing. But I do not totally understand what the need is of removing the assert. For GC, we have safepoints where we trigger GC or in other words, we mark code non-interruptible in which region GC won't trigger.

Also, could you add details on what test case was failing that you are trying to fix?

Reply:
The test case is `/home/qiao/tests/Linux.loongarch64.Release/JIT/Directed/StructABI/StructABI/StructABI.sh`.
And last month that was ok on LA64.

The messages are liking these:
```
{ Start Jitting Method    2 System.AppContext:Setup(ulong,ulong,int) (MethodHash=16ec513f) MinOpts
; Assembly listing for method System.AppContext:Setup(ulong,ulong,int)
; Emitting BLENDED_CODE for generic LOONGARCH64 CPU - Unix
; debuggable code
; fp based frame
; fully interruptible
; No PGO data
; Final local variable assignments
;
;  V00 arg0         [V00    ] (  1,  1   )    long  ->  [fp+48H]   do-not-enreg[]
;  V01 arg1         [V01    ] (  1,  1   )    long  ->  [fp+40H]   do-not-enreg[]
;  V02 arg2         [V02    ] (  1,  1   )     int  ->  [fp+3CH]   do-not-enreg[]
;  V03 loc0         [V03    ] (  1,  1   )     int  ->  [fp+38H]   do-not-enreg[]
;  V04 loc1         [V04    ] (  1,  1   )    bool  ->  [fp+34H]   do-not-enreg[]
;# V05 OutArgs      [V05    ] (  1,  1   )  lclBlk ( 0) [sp+00H]   do-not-enreg[] "OutgoingArgSpace"
;  V06 tmp1         [V06    ] (  1,  1   )     ref  ->  [fp+28H]   do-not-enreg[] must-init class-hnd exact "NewObj constructor temp"
;  V07 tmp2         [V07    ] (  1,  1   )     ref  ->  [fp+20H]   do-not-enreg[] must-init class-hnd "impSpillStackEnsure"
;  V08 tmp3         [V08    ] (  1,  1   )     ref  ->  [fp+18H]   do-not-enreg[] must-init class-hnd exact "impSpillStackEnsure"
;  V09 tmp4         [V09    ] (  1,  1   )     ref  ->  [fp+10H]   do-not-enreg[] must-init class-hnd exact "impSpillStackEnsure"
;
; Lcl frame size = 64

Thread 1 "corerun" hit Breakpoint 1, CodeGen::genSetRegToConst (this=0xaaaabdaa48, targetReg=REG_A1, targetType=TYP_REF, tree=0xaaaabdfc50)
    at /home/qiao/work_qiao/runtime/src/coreclr/jit/codegenloongarch64.cpp:1794
1794	                genSetRegToIcon(targetReg, cnsVal, targetType);
(gdb) p targetType
$13 = TYP_REF

### here will hit the error if calling the `genSetRegToIcon`.
### more info liking

(gdb) p compiler->opts.compReloc
$15 = false
(gdb) p con
$16 = (GenTreeIntConCommon *) 0xaaaabdfc50
(gdb) p compiler
$17 = (Compiler *) 0xaaaabd9ea8
(gdb) p con->ImmedValNeedsReloc(compiler)
$18 = false
(gdb) bt
#0  CodeGen::genSetRegToConst (this=0xaaaabdaa48, targetReg=REG_A1, targetType=TYP_REF, tree=0xaaaabdfc50)
    at /home/qiao/work_qiao/runtime/src/coreclr/jit/codegenloongarch64.cpp:1794
#1  0x000000ff75f1c284 in CodeGen::genCodeForTreeNode (this=0xaaaabdaa48, treeNode=0xaaaabdfc50)
    at /home/qiao/work_qiao/runtime/src/coreclr/jit/codegenloongarch64.cpp:5345
#2  0x000000ff75c7234c in CodeGen::genCodeForBBlist (this=0xaaaabdaa48) at /home/qiao/work_qiao/runtime/src/coreclr/jit/codegenlinear.cpp:470
#3  0x000000ff75c5bf90 in CodeGen::genGenerateMachineCode (this=0xaaaabdaa48) at /home/qiao/work_qiao/runtime/src/coreclr/jit/codegencommon.cpp:1917
#4  0x000000ff75c6e5b8 in CodeGenPhase::DoPhase() () at /home/qiao/work_qiao/runtime/src/coreclr/jit/gentree.h:1981
#5  0x000000ff75ea9d10 in Phase::Run (this=0xffffff2b10) at /home/qiao/work_qiao/runtime/src/coreclr/jit/phase.cpp:61
#6  0x000000ff75c69e40 in DoPhase(CodeGen*, Phases, void (CodeGen::*)()) () at /home/qiao/work_qiao/runtime/src/coreclr/jit/codegeninterface.h:322
#7  0x000000ff75c5b7c0 in CodeGen::genGenerateCode (this=0xaaaabdaa48, codePtr=0xffffff4050, nativeSizeOfCode=0xffffff4514)
    at /home/qiao/work_qiao/runtime/src/coreclr/jit/codegencommon.cpp:1692
#8  0x000000ff75c81e7c in Compiler::compCompile (this=0xaaaabd9ea8, methodCodePtr=0xffffff4050, methodCodeSize=0xffffff4514, compileFlags=0xffffff4078)
    at /home/qiao/work_qiao/runtime/src/coreclr/jit/compiler.cpp:4935

(gdb) p *con
$20 = {<GenTree> = {_vptr$GenTree = 0xff76195c28 <vtable for GenTreeIntCon+16>, gtOper = GT_CNS_INT, gtType = TYP_REF, gtOperSave = GT_NONE, gtCSEnum = 0 '\000', 
    gtLIRFlags = 0 '\000', gtAssertionInfo = {m_isNextEdgeAssertion = 0, m_assertionIndex = 0}, gtCostsInitialized = true, _gtCostEx = 1 '\001', _gtCostSz = 4 '\004', 
    gtRegTag = GenTree::GT_REGTAG_REG, _gtRegNum = 5 '\005', m_operandCount = 221 '\335', gtFlags = GTF_ICON_OBJ_HDL, gtDebugFlags = GTF_DEBUG_NODE_SMALL, gtVNPair = {
      m_liberal = 4294967295, m_conservative = 4294967295}, gtRsvdRegs = 0, gtNext = 0xaaaabe36d8, gtPrev = 0xaaaabe3660, gtTreeID = 78, gtSeqNum = 42, gtUseNum = 3, 
    static gtOperKindTable = 0xff7607e51c <GenTree::gtOperKindTable> "", static gtDebugOperKindTable = 0xff7607e598 <GenTree::gtDebugOperKindTable>, 
    static s_gtNodeSizes = 0xff761a6910 <GenTree::s_gtNodeSizes> 'x' <repeats 12 times>, "\330\330xxxx\330xxx\330xxxxx\330xx\330", 'x' <repeats 11 times>, "\330xxxx\330\330x\330", 'x' <repeats 30 times>, "\330x\330xxxxxxxxxx\330\330\330\330", 'x' <repeats 17 times>, "\330\330xxxxx", static s_vtablesForOpers = {0x0, 0x0, 0x0, 
      0xff76196a18 <vtable for GenTreeLclVar+16>, 0xff76196a78 <vtable for GenTreeLclFld+16>, 0xff76196a18 <vtable for GenTreeLclVar+16>, 0x0 <repeats 31 times>, 
      0xff76195d70 <vtable for GenTreeIndir+16>, 0x0 <repeats 35 times>, 0xff76195dd0 <vtable for GenTreeOp+16>, 0x0 <repeats 12 times>, 
      0xff76196e38 <vtable for GenTreeAddrMode+16>, 0x0, 0x0, 0x0, 0x0, 0xff76195dd0 <vtable for GenTreeOp+16>, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 
      0xff76196b08 <vtable for GenTreeCall+16>, 0x0 <repeats 23 times>}, static s_vtableForOp = 0xff76195dd0 <vtable for GenTreeOp+16>}, <No data fields>}
(gdb) p/x cnsVal
$22 = 0xff75768020

```

Reply:
> And at least that was ok on LA64 last month.

Do we understand what could have broken it?

Reply:
> > And at least that was ok on LA64 last month.
> 
> Do we understand what could have broken it?

I didn't look up it which commit resulted.
Here is just fixed it.

Reply:
Out of curiosity, do you run these tests periodically in your lab and find this out?

Reply:
> But I do not totally understand what the need is of removing the assert.

Also, did you try putting them? It is hard to relate arm64 code with loongarch.

Reply:
Because I was busy with other in september, I didn't test the CoreRoot durring the last 2~3 week.

Ok, I will try to find it.
And I will run CoreRoot periodically liking the last 2 months.

Reply:
Hi, @kunalspathak 
I had finded it #49576 which caused the error on LA64.

Reply:
Hi, @kunalspathak 
I think this PR is OK for LA64 which liking the #60228

Reply:
> Hi, @kunalspathak I think this PR is OK for LA64 which liking the #60228

I had deleted this function `genSetRegToIcon()` for LA64.